### PR TITLE
Handling cluster kubeconfig in zedrouter

### PR DESCRIPTION
**Pull request by zed-rishabh**
Publishing app inst metadata from zedrouter

Signed-off-by: Rishabh Gupta <rgupta@zededa.com>

nits

Signed-off-by: Rishabh Gupta <rgupta@zededa.com>

Fixing logs

Signed-off-by: Rishabh Gupta <rgupta@zededa.com>

### CODE REVIEW

File: pkg/pillar/cmd/zedrouter/server.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/server.go:350]: Isn't there a number of checks missing here such as checking it is a POST (and returning an error if not), checking that the content-type is json (and return error), and also check against some maximum size (to avoid OOM if an app sends a Gbyte of data).

File: pkg/pillar/types/kubeconfig.go
Comment (eriknordmark) [inline pkg/pillar/types/kubeconfig.go:8]: Should this definition include the AppInstMetadataType so that the same publication can be reused for future data flowing from app instance to the controller? 
If so it makes sense renamingh the structs etc to something like AppInstMetaData

File: pkg/pillar/types/zedroutertypes.go
Comment (eriknordmark) [inline pkg/pillar/types/zedroutertypes.go:2904]: This should be of type uuid.UUID and not string so we can have some type checking.

**Review by eriknordmark — COMMENTED** [review]
Can you please write up a page or two in the lf-edge wiki and post it to https://wiki.lfedge.org/display/EVE/Feature+Design+Proposals and send it to the eve mailing list?

That way we can get the API part, in particular the additional URLs, reviewed by the community.

**Review by eriknordmark — COMMENTED** [review]
Can you please write up a page or two in the lf-edge wiki and post it to https://wiki.lfedge.org/display/EVE/Feature+Design+Proposals and send it to the eve mailing list?

We should also find a place in docs/ to describe the API from the application's perspective.
Do we document the 169.254.169.254 for getting meta-data somewhere there?
That way we can get the API part, in particular the additional URLs, reviewed by the community.

**Comment by rvs** [comment]
Agreed with @eriknordmark -- this definitely needs a broader circulation on eve-tsc@ before it can go in.

File: pkg/pillar/types/kubeconfig.go
Comment (zed-rishabh) [reply pkg/pillar/types/kubeconfig.go:8]: AppInstMetaData structure is separately defined in zedroutertypes.go

File: pkg/pillar/cmd/zedrouter/server.go
Comment (zed-rishabh) [reply pkg/pillar/cmd/zedrouter/server.go:350]: Resolved.

File: pkg/pillar/types/zedroutertypes.go
Comment (zed-rishabh) [reply pkg/pillar/types/zedroutertypes.go:2904]: Resolved

**Comment by rvs** [comment]
Hey @zed-rishabh -- just a gentle ping -- do you plan to circulate your proposal this week?

File: pkg/pillar/cmd/zedrouter/server.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/server.go:384]: I made a comment over a week ago (don't know what happened to that comment) that it doesn't make any sense for the application to send the UUID. Insterad zedrouter/EVE should add the UUID. (Note that the other services use the source IP address to determine the appinstance using lookupAppNetworkStatusByAppIP.)

Comment (zed-rishabh) [reply pkg/pillar/cmd/zedrouter/server.go:384]: That changes are not committed yet. Waiting for the approval of the EVE community. Sent out proposal yesterday.

File: pkg/pillar/cmd/zedrouter/server.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/server.go:373]: I also made a comment/question a week and half ago asking whether the kube data is already json or whether this should support arbitrary bytes, plus asking for a more explicit check on the size limit.
My guess is that this implicitly verifies the size limit due to a truncated message resulting in invalid json, but I think we want an API which is just bytes, and an explicit size check hence no json decode here. But should we discuss in case I'm missing something.

(And I have no idea how github lost those two comments! Sigh)

Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/server.go]: Turns out those comments were in slack and not github.

File: pkg/pillar/cmd/zedrouter/server.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/server.go:363]: This is missing the size limit check. You should be able to use the pattern from utils which is
	content := make([]byte, maxReadSize)
	n, err := r.Read(content)
	if err != nil {
		err = fmt.Errorf("StatAndRead %s failed: %v", filename, err)
		if log != nil {
			log.Error(err)
		}
		return "", fi.ModTime(), err
	}
	if n == maxReadSize {
		err = fmt.Errorf("StatAndRead %s truncated after %d bytes",
			filename, maxReadSize)

File: pkg/pillar/cmd/zedrouter/server.go
Comment (eriknordmark) [inline pkg/pillar/cmd/zedrouter/server.go:357]: Maybe this should be changed or removed?

Comment (eriknordmark) [reply pkg/pillar/cmd/zedrouter/server.go:357]: Since it json we should keep it.

**Review by rvs — CHANGES_REQUESTED** [review]
I would like to request that we put this on hold until we sort out the issues raised on the mailing list/proposal page.

**Review by rvs — APPROVED** [review]

**Review by eriknordmark — APPROVED** [review]
Run Eden again

**Review by eriknordmark — APPROVED** [review]
Run eden again

**Pull request merged**
